cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -0,0 +1,883 @@
+
+  c_blst_generator = c_blst_p2_generator
+
+  c_blst_p_is_equal = c_blst_p2_is_equal
+  c_blst_p_is_inf = c_blst_p2_is_inf
+
+---- Safe Scalar types / marshalling
+
+sizeScalar :: forall i. Num i => i
+sizeScalar = fromIntegral c_size_blst_scalar
+
+newtype Scalar = Scalar (ForeignPtr Void)
+
+withIntScalar :: Integer -> (ScalarPtr -> IO a) -> IO a
+withIntScalar i go = do
+  s <- scalarFromInteger i
+  withScalar s go
+
+withScalar :: Scalar -> (ScalarPtr -> IO a) -> IO a
+withScalar (Scalar p2) go = do
+  withForeignPtr p2 (go . ScalarPtr)
+
+withNewScalar :: (ScalarPtr -> IO a) -> IO (a, Scalar)
+withNewScalar go = do
+  p2 <- mallocForeignPtrBytes sizeScalar
+  x <- withForeignPtr p2 (go . ScalarPtr)
+  return (x, Scalar p2)
+
+withNewScalar_ :: (ScalarPtr -> IO a) -> IO a
+withNewScalar_ = fmap fst . withNewScalar
+
+withNewScalar' :: (ScalarPtr -> IO a) -> IO Scalar
+withNewScalar' = fmap snd . withNewScalar
+
+cloneScalar :: Scalar -> IO Scalar
+cloneScalar (Scalar a) = do
+  b <- mallocForeignPtrBytes sizeScalar
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeScalar
+  return (Scalar b)
+
+sizeFr :: forall i. Num i => i
+sizeFr = fromIntegral c_size_blst_fr
+
+newtype Fr = Fr (ForeignPtr Void)
+
+withFr :: Fr -> (FrPtr -> IO a) -> IO a
+withFr (Fr p2) go = do
+  withForeignPtr p2 (go . FrPtr)
+
+withNewFr :: (FrPtr -> IO a) -> IO (a, Fr)
+withNewFr go = do
+  p2 <- mallocForeignPtrBytes sizeFr
+  x <- withForeignPtr p2 (go . FrPtr)
+  return (x, Fr p2)
+
+withNewFr_ :: (FrPtr -> IO a) -> IO a
+withNewFr_ = fmap fst . withNewFr
+
+withNewFr' :: (FrPtr -> IO a) -> IO Fr
+withNewFr' = fmap snd . withNewFr
+
+cloneFr :: Fr -> IO Fr
+cloneFr (Fr a) = do
+  b <- mallocForeignPtrBytes sizeFr
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeFr
+  return (Fr b)
+
+scalarToInteger :: Scalar -> IO Integer
+scalarToInteger scalar = withScalar scalar $ \scalarPtr -> do
+  allocaBytes sizeScalar $ \rawPtr -> do
+    c_blst_bendian_from_scalar rawPtr scalarPtr
+    cstrToInteger rawPtr sizeScalar
+
+cstrToInteger :: Ptr CChar -> Int -> IO Integer
+cstrToInteger p l = do
+  go l (castPtr p)
+  where
+    go :: Int -> Ptr CUChar -> IO Integer
+    go 0 _ = return 0
+    go n ptr = do
+      val <- peek ptr
+      res <- go (pred n) (plusPtr ptr 1)
+      return $ res .|. shiftL (fromIntegral val) (8 * pred n)
+
+integerToBS :: Integer -> ByteString
+integerToBS 0 = BS.empty
+integerToBS n
+  | n < 0
+  = error "Cannot convert negative Integer to ByteString"
+  | otherwise
+  = BS.snoc
+      (integerToBS (n `shiftR` 8))
+      (fromIntegral n)
+
+padBS :: Int -> ByteString -> ByteString
+padBS i b
+  | i > BS.length b
+  = BS.replicate (i - BS.length b) 0 <> b
+  | otherwise
+  = b
+
+integerAsCStr :: Integer -> (Ptr CChar -> Int -> IO a) -> IO a
+integerAsCStr n f = do
+  let bs = integerToBS n
+  BS.useAsCStringLen bs $ uncurry f
```

Comment:
This doesn't seem to be used anywhere (other than the tests, so maybe its just to check `cstrToInteger`), instead the padded version `integerAsCStrL` is used.

Reply:
Removed.